sdk/web-pubsub/web-pubsub-client/src/webPubSubClient.ts

```diff
@@ -0,0 +1,1121 @@
+    }
+    return this._groupMap.get(name) as WebPubSubGroup;
+  }
+
+  private ChangeState(newState: WebPubSubClientState): void {
+    console.debug(
+      `The client state transfer from ${this._state.toString()} to ${newState.toString()}`
+    );
+    this._state = newState;
+  }
+
+  private async operationExecuteWithRetry<T>(
+    inner: () => Promise<T>,
+    signal?: AbortSignalLike
+  ): Promise<T> {
+    let retryAttempt = 0;
+
+    while (true) {
+      try {
+        await inner.call(this);
+      } catch (err) {
+        retryAttempt++;
+        const delayInMs = this._messageRetryPolicy.nextRetryDelayInMs(retryAttempt);
+        if (delayInMs == null) {
+          throw err;
+        }
+
+        try {
+          await delay(delayInMs);
+        } catch {}
+
+        if (signal?.aborted) {
+          throw err;
+        }
+      }
+    }
+  }
+}
+
+class RetryPolicy {
+  private _retryOptions: WebPubSubRetryOptions;
+  private _maxRetriesToGetMaxDelay: number;
+
+  public constructor(retryOptions: WebPubSubRetryOptions) {
+    this._retryOptions = retryOptions;
+    this._maxRetriesToGetMaxDelay = Math.ceil(
+      Math.log2(this._retryOptions.maxRetryDelayInMs!) -
+        Math.log2(this._retryOptions.retryDelayInMs!) +
+        1
+    );
+  }
+
+  public nextRetryDelayInMs(retryAttempt: number): number | null {
+    if (retryAttempt > this._retryOptions.maxRetries!) {
+      return null;
+    } else {
+      if (this._retryOptions.mode! === "Fixed") {
+        return this._retryOptions.retryDelayInMs!;
+      } else {
+        return this.calculateExponentialDelay(retryAttempt);
+      }
+    }
+  }
+
+  private calculateExponentialDelay(attempt: number): number {
+    if (attempt >= this._maxRetriesToGetMaxDelay) {
+      return this._retryOptions.maxRetryDelayInMs!;
+    } else {
+      return (1 << (attempt - 1)) * this._retryOptions.retryDelayInMs!;
+    }
+  }
+}
+
+function delay(time: number): Promise<any> {
+  return new Promise((resolve) => setTimeout(resolve, time));
+}
+
+function sendAsync(socket: WebSocket, data: any, _?: AbortSignalLike): Promise<void> {
+  return new Promise((resolve, reject) => {
+    socket.send(data, (err) => {
+      if (err) {
+        reject(err);
+      } else {
+        resolve();
+      }
+    });
+  });
+}
+
+class WebPubSubGroup {
+  public readonly name: string;
+  public isJoined = false;
+
+  constructor(name: string) {
+    this.name = name;
+  }
+}
+
+class AckEntity {
+  private readonly _deferred: Deferred<WebPubSubResult>;
+
+  constructor(ackId: number) {
+    this._deferred = new Deferred();
+    this.ackId = ackId;
+  }
+
+  public ackId;
+
+  promise(): Promise<WebPubSubResult> {
+    return this._deferred.promise;
+  }
+
+  resolve(value: WebPubSubResult | PromiseLike<WebPubSubResult>): void {
+    this._deferred.resolve(value);
+  }
+
+  reject(reason?: any): void {
+    this._deferred.reject(reason);
+  }
+}
+
+class Deferred<T> {
```

Comment:
feels like overkill to have a `Deferred` class to use once in `AckEntity` - why not simply cache the below privates inside `AckEntity` itself?